pljava-so/src/main/c/BlackholeFDW.c

```diff
@@ -0,0 +1,708 @@
+{
+    // rel->serverid;
+    // rel->userid; // may be InvalidOid = current user)
+    // rel->useriscurrent
+    // rel->fdwroutine
+    // rel->fdw_private
+
+    // rel->rows
+    // rel->relid (only base rel, not joins)
+    // rel->min_attr  (often <0)
+    // rel->max_attr;
+
+    // ForeignTable *table = GetForeignTableExtended(foreigntableid, flags);  // expects relid?
+    // ForeignServer *server = GetForeignServerExtended(table->serverid, flags);
+
+    // ForeignTable *table = GetForeignTable(rel->relid);  // expects relid?
+    // ForeignServer *server = GetForeignServer(table->serverid);
+    // ForeignDataWrapper *fdw = GetForeignDataWrapper(server->fdwid);
+
+    // List *ftOptions = table->options;
+    // List *srvOptions = server->options;
+    // List *fdwOptions = fdw->options;
+    // List *userOptions = um->options;
+    // List *columnOptions = GetForeignColumnOptions(rel->relid, (AttrNumber) 0);
+
+    // macro
+    // char *username = MappingUserName(userid)
+
+    // Oid serverId = svr->serverid;
+    // Oid fdwid = svr->fdwid;
+    // Oid srvOwnerId = svr->ownerid;
+    // char *servername = svr->servername;
+    // char *serverType = svr->servertype; // optional
+    // char *serverVersion = svr->serverversion; // optional
+
+    // Oid fwdid = fdw->fdwid;
+    // Oid owner = fdw->owner;
+    // char *fdwname = fdw->fdwname;
+    // Oid fdwhandler = fdw->fdwhandler;
+    // Oid fdwvalidator = fdw->fdwvalidator;
+
+    // Oid usermappingoid = um->umid;
+    // Oid userId = um->userId;
+    // Oid umServerId = um->serverId;
+
+    // wrapper->options;
+    // wrapper->owner;
+}
+
+static void
+blackholeGetForeignRelSize(PlannerInfo *root,
+                           RelOptInfo *baserel,
+                           Oid foreigntableid) {
+    /*1
+     * Obtain relation size estimates for a foreign table. This is called at
+     * the beginning of planning for a query that scans a foreign table. root
+     * is the planner's global information about the query; baserel is the
+     * planner's information about this table; and foreigntableid is the
+     * pg_class OID of the foreign table. (foreigntableid could be obtained
+     * from the planner data structures, but it's passed explicitly to save
+     * effort.)
+     *
+     * This function should update baserel->rows to be the expected number of
+     * rows returned by the table scan, after accounting for the filtering
+     * done by the restriction quals. The initial value of baserel->rows is
+     * just a constant default estimate, which should be replaced if at all
+     * possible. The function may also choose to update baserel->width if it
+     * can compute a better estimate of the average result row width.
+     */
+
+    elog(NOTICE, "entering function %s", __func__);
+
+#ifdef USE_JAVA
+	// JAVA CONSTRUCTOR based on foreigntable id
+    baserel->fdw_private = JNI_getForeignRelSize(user, root, baserel);
```

Comment:
Function names starting with `JNI_` in PL/Java are `JNICalls.c` wrappers of Java JNI functions. That prefix shouldn't be used here.